Author: Luca Boccassi <luca.boccassi@gmail.com>
Description: Fix drm master device API usage for kernel >= 4.8
 From kernel 4.8 and newer (commits 34a839c68 5e84c2690b8 d6ed682eba5) DRM
 master device APIs have changed. Patch nvidia-drm kernel module code to fix
 build failure.
--- a/nvidia-drm/nvidia-drm-modeset.c
+++ b/nvidia-drm/nvidia-drm-modeset.c
@@ -37,6 +37,8 @@
 #include <drm/drm_atomic_helper.h>
 #include <drm/drm_crtc.h>
 
+#include <linux/version.h>
+
 /*
  * In kernel versions before the addition of
  * drm_crtc_state::connectors_changed, connector changes were
@@ -675,7 +677,11 @@
         goto failed;
     }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
     drm_atomic_helper_swap_state(dev, state);
+#else
+    drm_atomic_helper_swap_state(state, true);
+#endif
 
     nvidia_drm_update_head_mode_config(state, requested_config);
 
--- a/nvidia-drm/nvidia-drm-drv.c
+++ b/nvidia-drm/nvidia-drm-drv.c
@@ -43,6 +43,11 @@
 #include <drm/drm_gem.h>
 #endif
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0)
+#include <drm/drm_auth.h>
+#endif
+
 static struct nvidia_drm_device *dev_list = NULL;
 
 #if defined(NV_DRM_ATOMIC_MODESET_AVAILABLE)
@@ -419,7 +424,11 @@
 
 static
 void nvidia_drm_master_drop(struct drm_device *dev,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
                             struct drm_file *file_priv, bool from_release)
+#else
+                            struct drm_file *file_priv)
+#endif
 {
     struct nvidia_drm_device *nv_dev = dev->dev_private;
     int ret;
@@ -452,7 +461,11 @@
     mutex_lock(&dev->master_mutex);
 
     if (!file_priv->is_master ||
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
         !file_priv->minor->master)
+#else
+        !dev->master)
+#endif
     {
         goto done;
     }
@@ -473,7 +486,11 @@
      * NVKMS modeset ownership, because nvidia_drm_master_set()'s call to
      * grabOwnership() will fail.
      */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
     drm_master_put(&file_priv->minor->master);
+#else
+    drm_master_put(&dev->master);
+#endif
     file_priv->is_master = 0;
 
     ret = 0;
